### ISSUE TITLE
Cannot reference internal libraries with cabal-version: 3.4

### ISSUE THREAD

**Issue by noughtmare** [haskell/cabal#7565]
**Describe the bug**

When using `cabal-version: 3.4` it seems like I cannot directly reference internal libraries anymore. I can only use the experimental `package:internal-lib` notation.

**To Reproduce**

1. Make a cabal file with these contents:

    ```cabal
    cabal-version: 3.4
    name: my-lib
    version: 0.1.0.0

    library my-lib-internal

    library
      build-depends: my-lib-internal
    ```
2. `cabal v2-build`

**Expected behavior**

A successful build like if I use `cabal-version: 3.0`.

**System information**
 - Operating system: NixOS 21.05
 - `cabal`, `ghc` versions

    ```
    cabal-install version 3.4.0.0
    compiled using version 3.4.0.0 of the Cabal library 
    ```

    ```
    The Glorious Glasgow Haskell Compilation System, version 8.10.4
    ```

**Additional context**

From this stackoverflow question: https://stackoverflow.com/q/68880126/15207568

**Comment by fgaz** [comment]
This was partially discussed in #7286 and related tickets. Unfortunately removing the experimental label is blocked on #7270

**Comment by noughtmare** [comment]
I guess I'm still confused why cabal can't find the internal library without the `:` notation. Wouldn't it be sensible to default to internal libraries when resolving names that do not use that `:` notation, or at most showing a specific error if a package with the same name exists externally?

**Comment by kczulko** [comment]
Hi,

Is this still an issue? That's my result when I run `cabal build all` against [github.com/tweag/gazelle_cabal](https://github.com/tweag/gazelle_cabal/tree/main/example). It looks fine (I think so) for cabal 3.6.2.0. Maybe it can be closed? E.g. this entry doesn't cause a failure https://github.com/tweag/gazelle_cabal/blob/main/example/package-b/package-b.cabal#L37

CC @Mikolaj @fgaz 

```
 kczulko@workstation: ~/Projects/gazelle_cabal/example $ cabal --version
cabal-install version 3.6.2.0
compiled using version 3.6.2.0 of the Cabal library
 kczulko@workstation: ~/Projects/gazelle_cabal/example $ cabal build all
Build profile: -w ghc-8.10.7 -O1
In order, the following will be built (use -v for more details):
 - package-a-0.1.0.0 (lib) (first run)
 - package-a-0.1.0.0 (exe:package-a) (first run)
Warning: package-a.cabal:44:17: colon specifier is experimental feature (issue
#5660)
Warning: package-a.cabal:48:17: colon specifier is experimental feature (issue
#5660)
Warning: package-a.cabal:52:17: colon specifier is experimental feature (issue
#5660)
Warning: package-a.cabal:56:13: colon specifier is experimental feature (issue
#5660)
Warning: package-a.cabal:24:23: visibility is experimental feature (issue
#5660)
Warning: package-a.cabal:15:23: visibility is experimental feature (issue
#5660)
Configuring library for package-a-0.1.0.0..
Warning: Packages using 'cabal-version: >= 1.10' and before 'cabal-version:
3.4' must specify the 'default-language' field for each component (e.g.
Haskell98 or Haskell2010). If a component uses different languages in
different modules then list the other ones in the 'other-languages' field.
Preprocessing library for package-a-0.1.0.0..
Building library for package-a-0.1.0.0..
[1 of 4] Compiling PackageA.Other.E ( /home/kczulko/Projects/gazelle_cabal/example/dist-newstyle/build/x86_64-linux/ghc-8.10.7/package-a-0.1.0.0/build/PackageA/Other/E.hs, /home/kczulko/Projects/gazelle_cabal/example/dist-newstyle/build/x86_64-linux/ghc-8.10.7/package-a-0.1.0.0/build/PackageA/Other/E.o, /home/kczulko/Projects/gazelle_cabal/example/dist-newstyle/build/x86_64-linux/ghc-8.10.7/package-a-0.1.0.0/build/PackageA/Other/E.dyn_o )
[2 of 4] Compiling PackageA.Other.D ( src/PackageA/Other/D.lhs, /home/kczulko/Projects/gazelle_cabal/example/dist-newstyle/build/x86_64-linux/ghc-8.10.7/package-a-0.1.0.0/build/PackageA/Other/D.o, /home/kczulko/Projects/gazelle_cabal/example/dist-newstyle/build/x86_64-linux/ghc-8.10.7/package-a-0.1.0.0/build/PackageA/Other/D.dyn_o )
[3 of 4] Compiling PackageA.Other.B ( src/PackageA/Other/B.hs, /home/kczulko/Projects/gazelle_cabal/example/dist-newstyle/build/x86_64-linux/ghc-8.10.7/package-a-0.1.0.0/build/PackageA/Other/B.o, /home/kczulko/Projects/gazelle_cabal/example/dist-newstyle/build/x86_64-linux/ghc-8.10.7/package-a-0.1.0.0/build/PackageA/Other/B.dyn_o )
[4 of 4] Compiling PackageA.Exposed.A ( src/PackageA/Exposed/A.hs, /home/kczulko/Projects/gazelle_cabal/example/dist-newstyle/build/x86_64-linux/ghc-8.10.7/package-a-0.1.0.0/build/PackageA/Exposed/A.o, /home/kczulko/Projects/gazelle_cabal/example/dist-newstyle/build/x86_64-linux/ghc-8.10.7/package-a-0.1.0.0/build/PackageA/Exposed/A.dyn_o )
Warning: package-a.cabal:44:17: colon specifier is experimental feature (issue
#5660)
Warning: package-a.cabal:48:17: colon specifier is experimental feature (issue
#5660)
Warning: package-a.cabal:52:17: colon specifier is experimental feature (issue
#5660)
Warning: package-a.cabal:56:13: colon specifier is experimental feature (issue
#5660)
Warning: package-a.cabal:24:23: visibility is experimental feature (issue
#5660)
Warning: package-a.cabal:15:23: visibility is experimental feature (issue
#5660)
Configuring executable 'package-a' for package-a-0.1.0.0..
Warning: Packages using 'cabal-version: >= 1.10' and before 'cabal-version:
3.4' must specify the 'default-language' field for each component (e.g.
Haskell98 or Haskell2010). If a component uses different languages in
different modules then list the other ones in the 'other-languages' field.
Preprocessing executable 'package-a' for package-a-0.1.0.0..
Building executable 'package-a' for package-a-0.1.0.0..
[1 of 2] Compiling PackageA.Other.C ( app/PackageA/Other/C.hs, /home/kczulko/Projects/gazelle_cabal/example/dist-newstyle/build/x86_64-linux/ghc-8.10.7/package-a-0.1.0.0/x/package-a/build/package-a/package-a-tmp/PackageA/Other/C.o )
[2 of 2] Compiling Main             ( app/Main.hs, /home/kczulko/Projects/gazelle_cabal/example/dist-newstyle/build/x86_64-linux/ghc-8.10.7/package-a-0.1.0.0/x/package-a/build/package-a/package-a-tmp/Main.o )
Linking /home/kczulko/Projects/gazelle_cabal/example/dist-newstyle/build/x86_64-linux/ghc-8.10.7/package-a-0.1.0.0/x/package-a/build/package-a/package-a ...
```

**Comment by phadej** [comment]
See https://github.com/haskell/cabal/issues/6083

This issue is IMO wontfix, except for possibly better diagnostics from solver.

**Comment by kczulko** [comment]
Sorry, I just understood what I did (therefore I've removed my previous msg)... In this case `cabal-version` is the entry in the cabal file... So indeed, for `cabal-version` > 3.4 it is still reproducible. 

I'm just going through the issues related to internal libraries and I've found this one. Thx @phadej for explanation.

**Comment by jneira** [comment]
Not sure if the issue is actionable: internal libs can be referenced for cabal-version 3.4 and it is documented
Additional improvements could be discussed in ne issues.

**Comment by noughtmare** [comment]
Actions I would like to see happen:

* Update documentation such as [this section about internal libraries](https://cabal.readthedocs.io/en/3.8/cabal-package.html#sublibs) to use the now required `package:sublib` syntax. Maybe there are more places that still use the old syntax.
* Improve the error message in the case that an unknown package name happens to be the name of an internal library. Then the error message can suggest using the new syntax.

**Comment by jneira** [comment]
Many thanks for summarize needed changes in docs, it seems we have the sketch for a pull request updating documentation!
